File: harness_builder/core/spec.py
"""
HarnessSpec — the contract between the BUILDER (which generates harnesses from
a user prompt) and the RUNTIME (which executes them).

A harness is a directory:

    harnesses/deep_research/
    ├── harness.yaml        <- this spec
    ├── skills/*.md         <- procedural memory (how each agent should act)
    └── memory/             <- semantic + episodic stores (created at runtime)

Everything inside an agent run is ephemeral (see architecture diagram);
the spec + skills + memory stores are what persist.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path

import yaml

PATTERNS = ["pipeline", "fanout", "expert_pool", "producer_reviewer",
            "supervisor", "hierarchical"]


@dataclass
class AgentSpec:
    name: str
    role: str                                # one-line purpose
    system_prompt: str                       # full system prompt for this agent
    model: str = "anthropic/claude-sonnet-4-6"   # provider/model routing string
    tools: list[str] = field(default_factory=list)  # names from the tool registry
    skills: list[str] = field(default_factory=list) # skill files loaded into context
    output_format: str = ""   # anatomy §4: how to structure responses (~400 tok)
    max_turns: int = 12


@dataclass
class MemorySpec:
    semantic: bool = True          # vector-ish store of durable facts
    episodic: bool = True          # SQLite of dated events / past runs
    consolidate_after: int = 3     # summarizer agent runs after N new sessions
    summarizer_model: str = "anthropic/claude-haiku-4-5-20251001"  # cheaper model


@dataclass
class GuardrailSpec:
    max_total_tokens: int = 300_000
    max_wall_seconds: int = 900
    shell_deny_patterns: list[str] = field(default_factory=lambda: [
        r"\brm\s+-rf\s+/", r"\bsudo\b", r"\bmkfs\b", r"\bssh\b", r">\s*/dev/",
    ])


@dataclass
class LoopSpec:
    """The EXTERNAL loop (outside the harness): plan-driven execution."""
    planner_model: str = "anthropic/claude-sonnet-4-6"
    max_cycles: int = 12            # total harness invocations across the loop
    max_attempts_per_step: int = 2  # retries before replanning around a step
    replan_on_failure: bool = True
    step_verify: str = "judge"      # "judge" (LLM per-step gate) | "none"


@dataclass
class EvalSpec:
    quality_criteria: list[str] = field(default_factory=list)  # judged by LLM
    judge_model: str = "anthropic/claude-sonnet-4-6"
    pass_threshold: float = 7.0    # 0-10; gate for the ralph loop / release


@dataclass
class HarnessSpec:
    name: str
    description: str
    pattern: str                              # one of PATTERNS
    agents: list[AgentSpec]
    command: str = ""                         # friendly launcher name (e.g. youvid)
    accent: str = ""                          # TUI accent hex the architect picks
    commands: list = field(default_factory=list)  # [{name, description, task}]
    flow: list = field(default_factory=list)  # pattern-specific wiring (agent names)
    supervisor: str | None = None             # for supervisor/hierarchical patterns
    memory: MemorySpec = field(default_factory=MemorySpec)
    guardrails: GuardrailSpec = field(default_factory=GuardrailSpec)
    eval: EvalSpec = field(default_factory=EvalSpec)
    loop: LoopSpec = field(default_factory=LoopSpec)
    mcp_servers: list = field(default_factory=list)  # [{name, command, args}]

    # ------------------------------------------------------------------ io
    @staticmethod
    def load(path: str | Path) -> "HarnessSpec":
        path = Path(path)
        if path.is_dir():
            path = path / "harness.yaml"
        data = yaml.safe_load(path.read_text())
        return HarnessSpec.from_dict(data)

    @staticmethod
    def from_dict(d: dict) -> "HarnessSpec":
        agents = [AgentSpec(**a) for a in d.get("agents", [])]
        spec = HarnessSpec(
            name=d["name"], description=d.get("description", ""),
            pattern=d["pattern"], agents=agents,
            command=d.get("command", ""),
            accent=d.get("accent", "") or "",
            commands=d.get("commands", []) or [],
            flow=d.get("flow", []), supervisor=d.get("supervisor"),
            memory=MemorySpec(**d.get("memory", {})),
            guardrails=GuardrailSpec(**d.get("guardrails", {})),
            eval=EvalSpec(**d.get("eval", {})),
            loop=LoopSpec(**d.get("loop", {})),
            mcp_servers=d.get("mcp_servers", []),
        )
        spec.validate()
        return spec

    def to_dict(self) -> dict:
        from dataclasses import asdict
        return asdict(self)

    def save(self, harness_dir: str | Path) -> Path:
        d = Path(harness_dir)
        (d / "skills").mkdir(parents=True, exist_ok=True)
        (d / "memory").mkdir(exist_ok=True)
        out = d / "harness.yaml"
        out.write_text(yaml.safe_dump(self.to_dict(), sort_keys=False,
                                      allow_unicode=True, width=100))
        return out

    # ------------------------------------------------------------ checks
    def validate(self):
        if self.pattern not in PATTERNS:
            raise ValueError(f"pattern must be one of {PATTERNS}, got '{self.pattern}'")
        if self.command and not re.fullmatch(r"[a-z][a-z0-9_-]{0,23}", self.command):
            raise ValueError(f"command must be short lowercase (got '{self.command}')")
        if self.accent and not re.fullmatch(r"#[0-9a-fA-F]{6}", self.accent):
            raise ValueError(f"accent must be a #RRGGBB hex (got '{self.accent}')")
        for c in self.commands:
            if not isinstance(c, dict) or not c.get("name") or not c.get("task"):
                raise ValueError(f"each command needs name+task (got {c!r})")
            if not re.fullmatch(r"[a-z][a-z0-9_-]{0,23}", c["name"]):
                raise ValueError(f"command name must be short lowercase "
                                 f"(got '{c['name']}')")
        names = [a.name for a in self.agents]
        if len(names) != len(set(names)):
            raise ValueError(f"duplicate agent names: {names}")
        if self.pattern in ("supervisor", "hierarchical"):
            if not self.supervisor or self.supervisor not in names:
                raise ValueError(f"pattern '{self.pattern}' needs a valid 'supervisor' "
                                 f"agent name; got '{self.supervisor}'")
        flat = [x for item in self.flow
                for x in (item if isinstance(item, list) else [item])]
        unknown = [x for x in flat if x not in names]
        if unknown:
            raise ValueError(f"flow references unknown agents: {unknown}")

    def agent(self, name: str) -> AgentSpec:
        for a in self.agents:
            if a.name == name:
                return a
        raise KeyError(name)

File: harness-rs/src/spec.rs
//! HarnessSpec — deserializes the SAME harness.yaml the Python builder emits.
//! The YAML file is the contract; both runtimes execute identical harnesses.

use anyhow::{bail, Context, Result};
use serde::Deserialize;
use std::path::{Path, PathBuf};

pub const PATTERNS: [&str; 6] = ["pipeline", "fanout", "expert_pool",
    "producer_reviewer", "supervisor", "hierarchical"];

fn d_model() -> String { "anthropic/claude-sonnet-4-6".into() }
fn d_max_turns() -> u32 { 12 }
fn d_true() -> bool { true }
fn d_consolidate() -> u32 { 3 }
fn d_summarizer() -> String { "anthropic/claude-haiku-4-5-20251001".into() }
fn d_tokens() -> u64 { 300_000 }
fn d_wall() -> u64 { 900 }
fn d_deny() -> Vec<String> {
    vec![r"\brm\s+-rf\s+/".into(), r"\bsudo\b".into(), r"\bmkfs\b".into(),
         r"\bssh\b".into(), r">\s*/dev/".into()]
}
fn d_judge() -> String { d_model() }
fn d_threshold() -> f64 { 7.0 }

#[derive(Debug, Clone, Deserialize)]
pub struct AgentSpec {
    pub name: String,
    pub role: String,
    pub system_prompt: String,
    #[serde(default = "d_model")]
    pub model: String,
    #[serde(default)]
    pub tools: Vec<String>,
    #[serde(default)]
    pub skills: Vec<String>,
    #[serde(default = "d_max_turns")]
    pub max_turns: u32,
}

#[derive(Debug, Clone, Deserialize)]
pub struct MemorySpec {
    #[serde(default = "d_true")]
    pub semantic: bool,
    #[serde(default = "d_true")]
    pub episodic: bool,
    #[serde(default = "d_consolidate")]
    pub consolidate_after: u32,
    #[serde(default = "d_summarizer")]
    pub summarizer_model: String,
}
impl Default for MemorySpec {
    fn default() -> Self {
        Self { semantic: true, episodic: true, consolidate_after: 3,
               summarizer_model: d_summarizer() }
    }
}

#[derive(Debug, Clone, Deserialize)]
pub struct GuardrailSpec {
    #[serde(default = "d_tokens")]
    pub max_total_tokens: u64,
    #[serde(default = "d_wall")]
    pub max_wall_seconds: u64,
    #[serde(default = "d_deny")]
    pub shell_deny_patterns: Vec<String>,
}
impl Default for GuardrailSpec {
    fn default() -> Self {
        Self { max_total_tokens: d_tokens(), max_wall_seconds: d_wall(),
               shell_deny_patterns: d_deny() }
    }
}

#[derive(Debug, Clone, Deserialize)]
pub struct EvalSpec {
    #[serde(default)]
    pub quality_criteria: Vec<String>,
    #[serde(default = "d_judge")]
    pub judge_model: String,
    #[serde(default = "d_threshold")]
    pub pass_threshold: f64,
}
impl Default for EvalSpec {
    fn default() -> Self {
        Self { quality_criteria: vec![], judge_model: d_judge(),
               pass_threshold: d_threshold() }
    }
}

/// flow entries may be "name" or ["a","b"] (parallel groups)
#[derive(Debug, Clone, Deserialize)]
#[serde(untagged)]
pub enum FlowItem {
    One(String),
    Group(Vec<String>),
}

#[derive(Debug, Clone, Deserialize)]
pub struct HarnessSpec {
    pub name: String,
    #[serde(default)]
    pub description: String,
    pub pattern: String,
    pub agents: Vec<AgentSpec>,
    #[serde(default)]
    pub flow: Vec<FlowItem>,
    #[serde(default)]
    pub supervisor: Option<String>,
    #[serde(default)]
    pub memory: MemorySpec,
    #[serde(default)]
    pub guardrails: GuardrailSpec,
    #[serde(default, rename = "eval")]
    pub eval_spec: EvalSpec,
}

impl HarnessSpec {
    pub fn load(path: &Path) -> Result<(Self, PathBuf)> {
        let (file, dir) = if path.is_dir() {
            (path.join("harness.yaml"), path.to_path_buf())
        } else {
            (path.to_path_buf(),
             path.parent().unwrap_or(Path::new(".")).to_path_buf())
        };
        let text = std::fs::read_to_string(&file)
            .with_context(|| format!("reading {}", file.display()))?;
        let spec: HarnessSpec = serde_yaml::from_str(&text)
            .with_context(|| format!("parsing {}", file.display()))?;
        spec.validate()?;
        Ok((spec, dir))
    }

    pub fn validate(&self) -> Result<()> {
        if !PATTERNS.contains(&self.pattern.as_str()) {
            bail!("pattern must be one of {:?}, got '{}'", PATTERNS, self.pattern);
        }
        let names: Vec<&str> = self.agents.iter().map(|a| a.name.as_str()).collect();
        let mut dedup = names.clone();
        dedup.sort();
        dedup.dedup();
        if dedup.len() != names.len() {
            bail!("duplicate agent names: {:?}", names);
        }
        if matches!(self.pattern.as_str(), "supervisor" | "hierarchical") {
            match &self.supervisor {
                Some(s) if names.contains(&s.as_str()) => {}
                other => bail!("pattern '{}' needs a valid 'supervisor'; got {:?}",
                               self.pattern, other),
            }
        }
        for item in &self.flow {
            let refs: Vec<&String> = match item {
                FlowItem::One(s) => vec![s],
                FlowItem::Group(g) => g.iter().collect(),
            };
            for r in refs {
                if !names.contains(&r.as_str()) {
                    bail!("flow references unknown agent '{}'", r);
                }
            }
        }
        Ok(())
    }

    pub fn agent(&self, name: &str) -> Result<&AgentSpec> {
        self.agents.iter().find(|a| a.name == name)
            .with_context(|| format!("no agent named '{}'", name))
    }

    /// flattened execution order (flow if given, else all agents)
    pub fn flow_or_all(&self) -> Vec<String> {
        if self.flow.is_empty() {
            return self.agents.iter().map(|a| a.name.clone()).collect();
        }
        self.flow.iter().flat_map(|i| match i {
            FlowItem::One(s) => vec![s.clone()],
            FlowItem::Group(g) => g.clone(),
        }).collect()
    }
}
